tracer/src/Datadog.Trace/Ci/Ipc/CircularChannel.Receiver.cs

```diff
@@ -0,0 +1,188 @@
+            _disposed = 0;
+            _pollingTimer = new Timer(PollForMessages, null, TimeSpan.Zero, TimeSpan.FromMilliseconds(PollingInterval));
+        }
+
+        public event EventHandler<byte[]>? MessageReceived;
+
+        private void PollForMessages(object? state)
+        {
+            _pollingEventFinished.Reset();
+            try
+            {
+                var channel = _channel;
+                if (Interlocked.Read(ref _disposed) == 1 || Interlocked.Read(ref channel._disposed) == 1)
+                {
+                    Log.Error("CircularChannel: Channel is disposed. Cannot read data.");
+                    return;
+                }
+
+                if (MessageReceived is null)
+                {
+                    // To avoid loosing messages we stop the polling if there are no subscribers
+                    return;
+                }
+
+                try
+                {
+                    var hasHandle = channel._mutex.WaitOne(MutexTimeout);
+                    if (!hasHandle)
+                    {
+                        CIVisibility.Log.Error("CircularChannel: Failed to acquire mutex within the time limit.");
+                        return;
+                    }
+                }
+                catch (AbandonedMutexException ex)
+                {
+                    CIVisibility.Log.Error(ex, "CircularChannel: Mutex was abandoned.");
+                    return;
+                }
+                catch (ObjectDisposedException)
+                {
+                    // The mutex was disposed, nothing to do
+                    return;
+                }
+
+                object? messagesToHandle = null;
+                try
+                {
+                    using var accessor = channel._mmf.CreateViewAccessor();
+                    var writePos = accessor.ReadUInt16(0);
+                    var readPos = accessor.ReadUInt16(2);
+                    while (readPos != writePos)
+                    {
+                        var absoluteReadPos = HeaderSize + readPos;
+                        if (channel.BufferSize - absoluteReadPos < 2)
+                        {
+                            // Not space to read the length of the message, so we need to go back to 0
+                            readPos = 0;
+                            absoluteReadPos = HeaderSize;
+                        }
+
+                        var length = accessor.ReadUInt16(absoluteReadPos);
+
+                        // Simple sanity check
+                        if (length + 2 > channel.BufferBodySize)
+                        {
+                            // Handle error, reset pointers, or skip
+                            break;
+                        }
+
+                        var data = new byte[length];
+
+                        // Read the first part of the data
+                        var firstPartLength = Math.Min(length, channel.BufferSize - absoluteReadPos - 2);
```

Comment:
nit: I think you could get rid of all these `+2` `-2` if you updated `absoluteReadPos` after reading the length?

Reply:
kind of like it with the indexes.